mlflow/transformers.py

```diff
@@ -0,0 +1,851 @@
+    for component_name in component_types:
+        component = getattr(pipeline, component_name)
+        component.save_pretrained(root_path.joinpath(component_name))
+    if processor:
+        processor.save_pretrained(root_path.joinpath(_PROCESSOR_KEY))
+
+
+def _load_component(root_path: pathlib.Path, component_key: str, component_type):
+    """
+    Loads an individual component object from local disk.
+    """
+    import transformers
+
+    components_dir = root_path.joinpath(_COMPONENTS_BINARY_KEY)
+    component_path = components_dir.joinpath(component_key)
+    component_instance = getattr(transformers, component_type)
+    return component_instance.from_pretrained(component_path)
+
+
+def _generate_base_flavor_configuration(
+    model,
+    task: str,
+) -> Dict[str, str]:
+    """
+    Generates the base flavor metadata needed for reconstructing a pipeline from saved
+    components. This is important because the ``Pipeline`` class does not have a loader
+    functionality. The serialization of a Pipeline saves the model, configurations, and
+    metadata for ``FeatureExtractor``s, ``Processor``s, and ``Tokenizer``s exclusively.
+    This function extracts key information from the submitted model object so that the precise
+    instance types can be loaded correctly.
+    """
+
+    _validate_transformers_task_type(task)
+
+    flavor_configuration = {
+        _TASK_KEY: task,
+        _LOGGED_TYPE_KEY: _get_instance_type(model, True),
+        _INSTANCE_TYPE_KEY: _get_instance_type(model, False),
+        _MODEL_PATH_OR_NAME_KEY: _get_base_model_architecture(model),
+        _PIPELINE_MODEL_TYPE_KEY: _get_instance_type(model.model, False),
+    }
+
+    return flavor_configuration
+
+
+def _get_or_infer_task_type(model, task: Optional[str] = None) -> str:
+    """
+    Validates that a supplied task type is supported by the ``transformers`` library if supplied,
+    else, if not supplied, infers the appropriate task type based on the model type.
+    """
+    if task:
+        _validate_transformers_task_type(task)
+    else:
+        task = _infer_transformers_task_type(model)
+    return task
+
+
+def _infer_transformers_task_type(model) -> str:
+    """
+    Performs inference of the task type, used in generating a pipeline object based on the
+    underlying model's intended use case. This utility relies on the definitions within the
+    transformers pipeline construction utility functions.
+
+    :param model: Either the model or the Pipeline object that the task will be extracted or
+                  inferred from
+    :return: The task type string
+    """
+    from transformers import Pipeline
+    from transformers.pipelines import get_task
+
+    if isinstance(model, Pipeline):
+        return model.task
+    elif isinstance(model, dict):
+        # transformers provides guard conditions for potentially invalid entries
+        return get_task(model[_MODEL_KEY].name_or_path)
```

Comment:
It also seems that this can throw for certain model types, such as the dalle-mini one above:

```
In [1]:     from transformers.pipelines import get_task
   ...:
2023-03-27 22:45:19.553365: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
get_task(
In [2]: get_task("dalle-mini/vqgan_imagenet_f16_16384")
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-a7052311bddd> in <cell line: 1>()
----> 1 get_task("dalle-mini/vqgan_imagenet_f16_16384")

~/opt/anaconda3/envs/base2/lib/python3.8/site-packages/transformers/pipelines/__init__.py in get_task(model, use_auth_token)
    395         raise RuntimeError(f"Instantiating a pipeline without a task set raised an error: {e}")
    396     if not info.pipeline_tag:
--> 397         raise RuntimeError(
    398             f"The model {model} does not seem to have a correct `pipeline_tag` set to infer the task automatically"
    399         )

RuntimeError: The model dalle-mini/vqgan_imagenet_f16_16384 does not seem to have a correct `pipeline_tag` set to infer the task automatically
```

Can we catch such errors and re-throw a clearer MlflowException telling users to specify the `task` parameter in order to log the model?

Reply:
added logic to catch these and raised a clear direction for what to do to fix it